### PR TITLE
Adds support for handling interface changes to mdns behaviour.

### PR DESCRIPTION
**Pull request by dvc94ch**
Currently the mdns behaviour will panic if no adapter is available. To test on linux use the following commands:
```sh
sudo ifconfig wlan0 down
cargo run --example mdns-passive-discovery
sudo ifconfig wlan0 up
```

~Depends on https://github.com/paritytech/ip-watch/pull/1~

cc @Demi-Marie @rklaehn

### CODE REVIEW

**Comment by dvc94ch** [comment]
Updated ip-watch to work on windows too and tested that it handles changing networks.

File: protocols/mdns/src/service.rs
Comment (dvc94ch) [inline protocols/mdns/src/service.rs:224]: This is a little bit of a hack, but it works since `next` is called regularly.

File: protocols/mdns/src/service.rs
Comment (dvc94ch) [reply protocols/mdns/src/service.rs:150]: `#[cfg(unix)]`? That has been around forever, certainly since 1.0.0

Comment (rkuhn) [reply protocols/mdns/src/service.rs:150]: No, I mean the ability to attach attributes to statements — I certainly do remember that that didn’t work a year ago, which may have been the reason behind the previous approach with a nested function.

Comment (dvc94ch) [reply protocols/mdns/src/service.rs:150]: :facepalm: obviously that's what you meant. Don't know, looks like the rfc issue is still open https://github.com/rust-lang/rust/issues/15701

File: protocols/mdns/src/service.rs
Comment (dvc94ch) [reply protocols/mdns/src/service.rs]: ah yes, now that the prefix is checked properly, there shouldn't be any errors

File: protocols/mdns/src/service.rs
Comment (dvc94ch) [reply protocols/mdns/src/service.rs]: I guess leaving when the iface is down also makes sense

**Comment by dvc94ch** [comment]
Just noticed an issue, it returns routes on linux but interfaces on windows. Looks like @Demi-Marie intended to query routes? Unless there is a reason for that, I'll change the linux implementation of `if-watch` to query ip's.

File: protocols/mdns/src/service.rs
Comment (romanb) [reply protocols/mdns/src/service.rs]: What effect does `leave_multicast_v4` have when used with the address of an interface that is down? Naively I would expect it to be either a no-op or an error and as such `leave_multicast_v4` to be only useful when one wants to leave a multicast group on an interface that is still up, but I'm not too familiar with IP multicast configuration.

Comment (dvc94ch) [reply protocols/mdns/src/service.rs]: That's what I'd expect too: it to be a noop.

Comment (dvc94ch) [reply protocols/mdns/src/service.rs]: It seems that `join_multicast_v4` might be a noop on linux too. removing the code works too.

File: protocols/mdns/src/service.rs
Comment (romanb) [inline protocols/mdns/src/service.rs:209]: Could we perhaps better integrate the `if_watch.next()` future into the [select block that waits for the next inbound datagram or the next query interval](https://github.com/libp2p/rust-libp2p/blob/36c997c7e305baabd714c5d5a571ed71bc58ec13/protocols/mdns/src/service.rs#L273-L281), e.g. by combining it in another `select()` with `query_interval.next()`? It would avoid potential unnecessary delays in receiving `IfEvent::Up` while waiting for the query interval to elapse and there is no inbound traffic.

File: protocols/mdns/src/service.rs
Comment (romanb) [inline protocols/mdns/src/service.rs:230]: As mentioned in an earlier comment, I think it would be good to get clarity on whether doing this on an interface that is down makes any sense. If this is just a no-op in this situation, there is no reason to do it and if it is an error it causes unnecessary noise in the logs. If you're sure that doing this on an interface that is down is useful and only errors when there is a serious problem other than the interface being down, I'm fine with it.

**Review by romanb — COMMENTED** [review]
> Just noticed an issue, it returns routes on linux but interfaces on windows. Looks like @Demi-Marie intended to query routes? Unless there is a reason for that, I'll change the linux implementation of if-watch to query ip's. This means that the prefix is 32 on linux and the network prefix on windows.

Since I'm neither familiar with the details of `ip-watch` nor `if-watch`, could you elaborate on the observed difference and the changes you made? I'm generally just assuming that these libraries "do the right thing" for the purpose of network interface monitoring as desired for MDNS.

**Comment by romanb** [comment]
@mxinden If you could also take a quick look at these changes, since you've worked on this code before, it would be much appreciated.

**Comment by dvc94ch** [comment]
They do the right thing now. I switched the rtnetlink code to use `RTM_NEWADDR, RTM_DELADDR, RTM_GETADDR` instead of `RTM_NEWROUTE, RTM_DELROUTE, RTM_GETROUTE`. The result is now the same as running `ip address` on linux.

- https://man7.org/linux/man-pages/man7/rtnetlink.7.html

File: protocols/mdns/src/service.rs
Comment (dvc94ch) [reply protocols/mdns/src/service.rs:230]: Well it doesn't cause an error on linux or windows. I'll try to investigate more deeply especially on linux, but we are supporting any platform that rust compiles to I guess.

**Comment by romanb** [comment]
> They do the right thing now. I switched the rtnetlink code to use RTM_NEWADDR, RTM_DELADDR, RTM_GETADDR instead of RTM_NEWROUTE, RTM_DELROUTE, RTM_GETROUTE. The result is now the same as running ip address on linux.

Thanks for the link. There are some things about the implementation of `if-watch` that confuse me a bit:

  1. It is sending an [RTM_GETADDR](https://github.com/dvc94ch/if-watch/blob/b67232953ecfdb8a88485684ee942159556aaf25/src/unix/linux.rs#L83) request with an [rtmsg structure](https://github.com/dvc94ch/if-watch/blob/b67232953ecfdb8a88485684ee942159556aaf25/src/unix/linux.rs#L88), which according to the man pages is only applicable to `RTM_NEWROUTE`, `RTM_DELROUTE` and `RTM_GETROUTE` types. The expected payload for `RTM_GETADDR` et al on the other hand according to the manpage is an `ifaddrmsg` structure. I guess the `rtmsg` is simply ignored when sent with an `RTM_GETADDR` request?

  2. I only see `send_getaddr()` being called [once when initialising the watcher](https://github.com/dvc94ch/if-watch/blob/b67232953ecfdb8a88485684ee942159556aaf25/src/unix.rs#L77). How will the watcher ever receive further updates? The previous use of `RTM_GETROUTE` may have relied upon the `RTM_F_NOTIFY` flag, but that does not seem applicable to `RTM_GETADDR` because it is a flag for the `rtmsg` structure.

 3. Would `RTM_GETLINK` not be the most appropriate to list network interfaces? Of course, it may have the same problem as `RTM_GETADDR` in that it needs to be sent periodically, as far as I currently understand.

This is the first time I'm looking at `rtnetlink`, so I may be missing the obvious.

**Comment by dvc94ch** [comment]
It only needs to `send_getaddr` at the beginning to get the current interfaces. We subscribe to notifications by setting the `nl_groups` to `RTMGRP_IPV4_IFADDR` and `RTMGRP_IPV6_IFADDR`. I'm not sure about the `rtmsg` as I didn't find any documentation on it. The behaviour I'm observing is that it ignores the fields, but from looking at examples written in C, it needs to send at least the family as a payload. As the first element is a family of size u8, I'm not sure if the rest is just zeroes as common practice in C or if it uses a different struct.

File: protocols/mdns/src/service.rs
Comment (dvc94ch) [reply protocols/mdns/src/service.rs]: Done

**Review by mxinden — COMMENTED** [review]
The direct changes in this pull request look good to me. Thanks for making `libp2p-mdns` more robust. On the side I am very much in favor of getting rid of the async-std/tokio macros.

I want to highlight once more that this switches `libp2p-mdns` from previously using either tokio or async-std to [`async-io`](https://github.com/stjepang/async-io) wrapping an std udp socket. In case there is anyone objecting to this change, please raise your concerns here.

I still have to take a deeper look at the introduced `if-watch` dependency.

**Comment by dvc94ch** [comment]
Some maintainance stuff that might be good to do before merging is replacing the old net2 crate with socket2 as used in the tcp transport and remove the wasm-time dependency in favor of async-io, as mdns doesn't run in browsers anyway.

**Comment by dvc94ch** [comment]
Can we get this PR in soon? @romanb @mxinden

**Comment by mxinden** [comment]
As far as I can tell the current `libp2p-mdns` implementation on e.g. `master` works on MacOS. I don't have a device to test this assumption so please correct me if I am wrong. The `if-watch` crate does not [compile on MacOS](https://github.com/dvc94ch/if-watch/issues/1). Is dropping support for MacOS worth the reduced interface-change-detection latency we get through `if-watch`? If I am not mistaken we could as well retrieve the current interfaces on an interval instead, e.g. via the `if-addrs` crate (or the large `pnet` crate).

> Can we get this PR in soon? @romanb @mxinden

I appreciate your work, fixing the current `libp2p-mdns` implementation. I don't have much expertise interacting with the kernel on the level that `if-watch` does. Thus I can't judge whether the `panic`s and `unsafe`-usages e.g with `U32AlignedBuffer`  in`if-watch` are fine. In addition, `rust-libp2p` seems to be the very first user of the crate, right? I can spend a couple more hours this week reviewing the crate.

That is not to say that I am blocking this pull request. In case there is an approval by another maintainer I am fine with merging.

**Comment by dvc94ch** [comment]
Well `if-watch` was written for libp2p, in particular for mdns and tcp/quic transports, so it is the first crate. I can add an interval based solution for macos until someone adds proper support. The windows implementation isn't that great either, someone more into windows could probably improve that too.

**Comment by romanb** [comment]
The diff itself looks good to me, but as @mxinden said, even if we don't have dedicated tests or CI checks for it at the moment, in principle rust-libp2p certainly has the ambition to support MacOS, hence including a library that is known to lack support is not really an option. Whenever we get tickets here that rust-libp2p does not compile for some platform, especially when it is as common as MacOS, we necessarily need to look into it. Without access to the right system for testing (I don't have or use MacOS either) we still need to make a best-effort and deal with issues in collaboration with users that do use these platforms as they come up. So we need an approach here (via `if-watch` or otherwise) that is at least intended to work on MacOS as well, even if it may turn out to have issues that need to be resolved afterwards.

File: protocols/mdns/src/service.rs
Comment (romanb) [reply protocols/mdns/src/service.rs]: > I'll try to investigate more deeply especially on linux, [..]

Just curious, did you come to a conclusion on the matter?

**Comment by dvc94ch** [comment]
Fixed in 0.1.6, now has a fallback that polls in 10s intervals.

**Review by romanb — APPROVED** [review]

**Comment by romanb** [comment]
@dvc94ch You don't need to keep this branch mergeable, e.g. by merging master and resolving conflicts. We typically resolve conflicts and update versions and changelogs just before, respectively after merging. The only thing that is advisable is to eventually stop rebasing / force-pushing to PR branches, in order not to cause unnecessary problems when we want to make final adjustments to the branch before merging.

**Comment by mxinden** [comment]
Thanks for adding a fallback mode @dvc94ch.

@romanb feel free to go ahead with merging.

**Comment by romanb** [comment]
@mxinden Thanks for your feedback. I intend to merge and probably release a new version of `libp2p-mdns` tomorrow.

**Comment by dvc94ch** [comment]
any thoughts on the tcp pr so far? I didn't find a good way to spawn the if_task in an executor agnostik way, the `tokio` and `async-std` people do not want any solution that requires more than one allocation in total and has been accepted in an rfc process [0], so I'm just spawning a thread.

- [0] https://github.com/async-rs/async-std/issues/912

**Comment by romanb** [comment]
> any thoughts on the tcp pr so far?

I didn't get a chance to look at it yet in any detail but intend to do so by next week.

**Comment by eskimor** [comment]
Thanks @dvc94ch ! Looks really awesome, especially that we react to interface changes. One thing that struck me when skimming through the changes: Just joining the multicast group on all interfaces is enough to make sure, messages go out on each interface?! I mean it kinda makes sense, but my research indicated otherwise. Anyhow, I really like that PR!

**Comment by dvc94ch** [comment]
Can you elaborate on what your research indicates is required? It has worked for me on linux and windows using the following test procedure:
1. start the mdns example on linux and on windows
2. wait for both to discover each other
3. change network to mobile access point on both devices
4. wait for both to discover each other
I don't know about multiple interfaces

**Comment by eskimor** [comment]
I see, different use case! In my case I really had multiple interfaces at the same time and according to some man page (would have to dig that one up again), when binding via multicast address or 0.0.0.0 you start receiving on all interfaces, but outgoing messages will be transmitted to the default interface (which is usually the one with the default gw with the highest priority) - in any case that matched my experience as well so far.

I will retry with your implementation and will report any issues :-)

**Comment by eskimor** [comment]
In any case one can change the outgoing interface, so in my implementation I just iterated over all interfaces, changed the outgoing interface to the current one and sent out the message on each one.

https://github.com/libp2p/rust-libp2p/pull/1750/files#diff-d82ec3ce9fb38291708105fbda3b4caeb5aa4d2d44a59b27c1e5a5679f9b9badR235

**Comment by eskimor** [comment]
Actually memory served me wrong. There is a platform specific call for sending, which allows specifying the interface that should be used.

**Comment by dvc94ch** [comment]
so it doesn't work? can you open a PR for it, I assume the changes should be pretty small now?

**Comment by eskimor** [comment]
Well, considering that you have not tested that particular case, I am pretty sure it won't, but I will test it and will provide a PR if that assumption is correct. The changes should be rather minimal. The only issue is that it depends on platform specific functionality, so this complicates matters. MacOS might be fine, but I have no idea about Windows.

**Comment by dvc94ch** [comment]
My devices only have a wifi adapter. I can help test that it doesn't break anything. Feel free to cc me on the PR.

**Comment by eskimor** [comment]
Thanks @dvc94ch !

**Pull request merged**
